src/async-interleave-generator.mjs

```diff
@@ -0,0 +1,43 @@
+import { ensureAsyncIterable } from './internal/async-iterable'
+import raceTo from './internal/race-to'
+import map from './map'
+
+class AsyncInterleaveBuffer {
+  constructor (iterator) {
+    this._iterator = iterator
+    this._nextPromise = iterator.next()
+  }
+
+  async take () {
+    const next = await this._nextPromise
+    if (next.done) {
+      return undefined
+    }
+
+    this._nextPromise = this._iterator.next()
+    return next.value
+  }
+
+  async canTake () {
+    return (await this._nextPromise).done ? null : this
+  }
+}
+
+export default function interleaveGenerator (generatorFn) {
+  return (...iterables) => {
+    return (async function * () {
+      const iterators = map(iterable => ensureAsyncIterable(iterable)[Symbol.asyncIterator](), iterables)
+      const buffers = Array.from(map(iterator => new AsyncInterleaveBuffer(iterator), iterators))
```

Comment:
Why don't you just use `Array.prototype.map`? That would make the code faster, more concise, and lighter

```javascript
const buffers = iterables
  .map(iterable => ensureAsyncIterable(iterable)[Symbol.asyncIterator]())
  .map(iterator => new AsyncInterleaveBuffer(iterator))
```